cliqz-logger.js

```diff
@@ -0,0 +1,419 @@
+    // Here is an example of how the import might look like;
+    // EOL/*LS-12345*/var { MyDefaultLogger } = ChromeUtils.import("resource://gre/modules/CliqzLogger.jsm");EOL
+    // var __logger_var__12 = MyDefaultLogger.init("<some-file-path>","<some-file-name>");/*LE-12345*/EOL
+    buffer.unshift(
+      `${EOL}/*LS-${randomNumber}*/` +
+      `var { ${config.loggerName} } = ChromeUtils.import(` +
+        `"${config.loggerPath}"` +
+      `);${EOL}`,
+      `var ${loggerVarName} = ${config.loggerName}.init("${path}","${pathObject.name}");` +
+      `/*LE-${randomNumber}*/${EOL}`
+    );
+  } else if (importAs == 'es6') {
+    // For EcmaScript6 compliant version we use its' standard import;
+    // Here is an example of how the import might look like;
+    // EOL/*LS-12345*/import { MyDefaultLogger } from "resource://gre/modules/CliqzLogger.jsm";EOL
+    // var __logger_var__12 = MyDefaultLogger.init("<some-file-path>","<some-file-name>");/*LE-12345*/EOL
+    buffer.unshift(
+      `${EOL}/*LS-${randomNumber}*/` +
+      `import { ${config.loggerName} } from '${config.loggerPath}';${EOL}`,
+      `var ${loggerVarName} = ${config.loggerName}.init('${path}','${pathObject.name}');` +
+      `/*LE-${randomNumber}*/${EOL}`
+    );
+  }
+};
+
+// This function is responsible for injecting log function invocations into the path;
+const injectLogFunction = function(path, config, loggerVarName) {
+  // First let it resolve path to make it cross-platform;
+  const resolvedPath = systemPath.resolve(PWD, path);
+  readDataFromPath(resolvedPath, function(error, buffer) {
+    if (error != null) {
+      writeToConsole(
+        `${EOL}Some error happened while reading the path "${resolvedPath}"`,
+        `${EOL}${error}${EOL}`
+      );
+
+      return;
+    }
+
+    // We parse the path with nodejs systemPath to retrieve the name out of it;
+    const pathObject = systemPath.parse(path);
+    // Import a logging provider module;
+    injectLogger(buffer, config, path, pathObject, loggerVarName);
+
+    let source = buffer.join("");
+    // For now we inject log function invocations only into each if-clause we have found;
+    let data = findConditionalClauses(source);
+
+    source = source.split("");
+    for (let i = 0, l = data.length; i < l; i++) {
+      const curly = source[data[i].bodyStartsAt];
+
+      // Inject right after the opening curly brace of if-clause statement;
+      // We provide information about a line number of if-clause we are dealing with
+      // and its' predicate as a string as well;
+      source[data[i].bodyStartsAt] = curly +
+        `${EOL}${loggerVarName}(` +
+          `"${data[i].lineNumber + i}", "${data[i].predicate}"` +
+        `);"${loggerVarName}";`;
+    }
+    source = source.join("");
+    // After injection is completed we write sources back to resolvedPath;
+    writeDataToPath(resolvedPath, source);
+  });
+};
+
+const readDataFromPath = function(path, cb) {
```

Comment:
Why implementing it as a stream and not `fs.readFile` or `fs.readFileSync`? Eventually you will be joining all batches anyway.

Reply:
Because using stream always takes less memory to read the file.